processing/src/main/java/org/apache/druid/query/operator/GlueingPartitioningOperator.java

```diff
@@ -0,0 +1,279 @@
+        firstPartitionHandled = true;
+        int start = boundaries[currentIndex];
+        int end = boundaries[currentIndex + 1];
+        LimitedRowsAndColumns limitedRAC = new LimitedRowsAndColumns(rac, start, end);
+
+        if (isGlueingNeeded(previousRac, limitedRAC)) {
+          RowsAndColumns gluedRAC = getConcatRacForFirstPartition(previousRac, limitedRAC);
+          ensureMaxRowsMaterializedConstraint(gluedRAC.numRows());
+          previousRac = null;
+          currentIndex++;
+          return gluedRAC;
+        } else {
+          if (previousRac != null) {
+            RowsAndColumns temp = previousRac;
+            previousRac = null;
+            return temp;
+          }
+        }
+      }
+
+      int start = boundaries[currentIndex];
+      int end = boundaries[currentIndex + 1];
+      currentIndex++;
+      return new LimitedRowsAndColumns(rac, start, end);
+    }
+
+    /**
+     * Determines whether gluing is needed between the previous RowsAndColumns and the first partition
+     * of the current RowsAndColumns based on the partition columns. If the columns match, the two RACs
+     * will be glued together.
+     *
+     * @param previousRac The previous RowsAndColumns.
+     * @param firstPartitionOfCurrentRac The first partition of the current RowsAndColumns.
+     * @return true if gluing is needed, false otherwise.
+     */
+    private boolean isGlueingNeeded(RowsAndColumns previousRac, RowsAndColumns firstPartitionOfCurrentRac)
+    {
+      if (previousRac == null) {
+        return false;
+      }
+
+      final ConcatRowsAndColumns concatRac = getConcatRacForFirstPartition(previousRac, firstPartitionOfCurrentRac);
+      for (String column : partitionColumns) {
+        final Column theCol = concatRac.findColumn(column);
+        if (theCol == null) {
+          continue;
+        }
+        final ColumnAccessor accessor = theCol.toAccessor();
+        // Compare 1st row of previousRac and firstPartitionOfCurrentRac in [previousRac, firstPartitionOfCurrentRac] form.
+        int comparison = accessor.compareRows(0, previousRac.numRows());
+        if (comparison != 0) {
+          return false;
+        }
+      }
+      return true;
+    }
+
+    private ConcatRowsAndColumns getConcatRacForFirstPartition(RowsAndColumns previousRac, RowsAndColumns firstPartitionOfCurrentRac)
+    {
+      return new ConcatRowsAndColumns(new ArrayList<>(Arrays.asList(previousRac, firstPartitionOfCurrentRac)));
+    }
+  }
+
+  private void ensureMaxRowsMaterializedConstraint(int numRows)
+  {
+    if (maxRowsMaterialized != MAX_ROWS_MATERIALIZED_NO_LIMIT && numRows > maxRowsMaterialized) {
+      throw InvalidInput.exception(
+          "Too many rows to process (requested = %d, max = %d).",
+          numRows,
+          maxRowsMaterialized
+      );
+    }
+  }
+
+  @Override
+  protected Iterator<RowsAndColumns> getIteratorForRAC(RowsAndColumns rac)
+  {
+    return new GluedRACsIterator(rac);
```

Comment:
why not build the 1st glued part of the rac here - and reduce the complexity in the `next()` ?
that way it would be possible to make the Iterator class static

Reply:
Right now, the GluedRACsIterator class encapsulates all the glueing logic. If we start doing the glueing outside, then it defeats the point of having an iterator class in my opinion.

If we start doing the glueing in getIteratorForRAC(), then GluedRACsIterator would not have anything left to do - so we might as well not have any such class?

Maybe I'm misinterpreting what you have in mind. Thoughts?

Reply:
I think we're on the same page wrt this (as per the latest comments), so marking this resolved.